tests/functions.sh

```diff
@@ -243,12 +360,326 @@ function delete_nic {
+    but_id=$1
+    nic_id=$2
+    vni=$3
+    protocol=$4
+    port=$5
+    f=/tmp/butterfly-client.req
+    echo "add nic $nic_id with only port $port opened in butterfly $but_id in vni $vni"
+
+    echo -e "messages {
+  revision: 0
+  message_0 {
+    request {
+      nic_add {
+        id: \"nic-$nic_id\"
+        mac: \"52:54:00:12:34:0$nic_id\"
+        vni: $vni
+        ip: \"42.0.0.$nic_id\"
+        ip_anti_spoof: true
+        security_group: \"sg-1\"
+      }
+    }
+  }
+}
+messages {
+  revision: 0
+  message_0 {
+    request {
+      sg_add {
+        id: \"sg-1\"
+        rule {
+          direction: INBOUND
+          protocol: $protocol
+          port_start: $port
+          port_end: $port
+          cidr {
+            address: \"0.0.0.0\"
+            mask_size: 0
+          }
+        }
+      }
+    }
+  }
+}
+" > $f
+    request $but_id $nic_id $f
+
+    if ! test -e /tmp/qemu-vhost-nic-$nic_id ; then
+        echo "client failed: we should have a socket in /tmp/qemu-vhost-nic-$nic_id"
+        clean_all
+        exit 1
+    fi
+}
+
+function add_nic_no_rules {
+    but_id=$1
+    nic_id=$2
+    vni=$3
+    echo "add nic $nic_id no security group rules in butterfly $but_id"
+    f=/tmp/butterfly-client.req
+    add_nic_void $but_id $nic_id $vni $f
+    update_nic_sg $but_id $nic_id "sg-1"
+}
+
+function add_nic_no_sg {
+    but_id=$1
+    nic_id=$2
+    vni=$3
+    echo "add nic $nic_id no security group in butterfly $but_id in vni $vni"
+    f=/tmp/butterfly-client.req
+    add_nic_void $but_id $nic_id $vni $f
+}
+
+function update_sg_rule_full_open {
+    but_id=$1
+    nic_id=$2
+    sg=$3
+    echo "update nic $nic_id rules full open in butterfly $but_id"
+    f=/tmp/butterfly-client.req
+
+    echo -e "messages {
+  revision: 0
+  message_0 {
+    request {
+      nic_update {
+        id: \"nic-$nic_id\"
+        ip: \"42.0.0.$nic_id\"
+        ip_anti_spoof: true
+        security_group: \"$sg\"
+      }
+    }
+  }
+}
+messages {
+  revision: 0
+  message_0 {
+    request {
+      sg_add {
+        id: \"sg-1\"
+        rule {
+          direction: INBOUND
+          protocol: -1
+          cidr {
+            address: \"0.0.0.0\"
+            mask_size: 0
+          }
+        }
+      }
+    }
+  }
+}
+" > $f
+    request $but_id $nic_id $f
+
+    if ! test -e /tmp/qemu-vhost-nic-$nic_id ; then
+        echo "client failed: we should have a socket in /tmp/qemu-vhost-nic-$nic_id"
+        clean_all
+        exit 1
+    fi
+}
+
+function update_sg_rule_port_open {
+    but_id=$1
+    nic_id=$2
+    protocol=$3
+    port=$4
+    sg=$5
+    f=/tmp/butterfly-client.req
+    echo "update nic $nic_id with only port $port opened in butterfly $but_id in vni $vni"
+
+    echo -e "messages {
+  revision: 0
+  message_0 {
+    request {
+      nic_update {
+        id: \"nic-$nic_id\"
+        ip: \"42.0.0.$nic_id\"
+        ip_anti_spoof: true
+        security_group: \"$sg\"
+      }
+    }
+  }
+}
```

Comment:
Do you need to update nic's SG here ? This function just add a new rule in the SG.
If the NIC does not already use a SG, this should be done with update_nic_sg